vllm/engine/multiprocessing/mp_llm_engine.py

```diff
@@ -0,0 +1,259 @@
+        elif self.worker_use_ray:
+            engine_class = ray.remote(num_cpus=0)(self._engine_class).remote
+        else:
+            raise NotImplementedError("Not supported yet!")
+        return engine_class(*args, **kwargs)
+
+    def run_background_loop(self):
+        """Entrypoint that kicks off the background processing loop."""
+
+        # Allow RPCClient to query data in startup phase.
+        self.run_startup_loop()
+
+        # Kick off core processing loop.
+        self.run_engine_loop()
+
+    @contextmanager
+    def make_data_socket(
+            self) -> Iterator[zmq.Socket]:  # type: ignore[name-defined]
+        socket = self.ctx.socket(zmq.constants.ROUTER)
+        try:
+            socket.bind(self.data_ipc_path)
+            yield socket
+        finally:
+            socket.close(linger=0)
+
+    def run_startup_loop(self) -> None:
+        """Loop over startup RPCStatupRequest from RPCClient."""
+
+        with self.make_data_socket() as socket:
+
+            # Loop until the RPCClient has all the data it needs.
+            client_is_ready = False
+            while not client_is_ready:
+                try:
+                    identity, message = socket.recv_multipart(copy=False)
+                    request: RPCStartupRequest = pickle.loads(message.buffer)
+
+                    # Handle the query from the Client.
+                    if request == RPCStartupRequest.GET_MODEL_CONFIG:
+                        response = self.engine.get_model_config()
+                    elif request == RPCStartupRequest.GET_DECODING_CONFIG:
+                        response = self.engine.get_decoding_config()
+                    elif request == RPCStartupRequest.GET_LORA_CONFIG:
+                        response = self.engine.get_lora_config()
+                    elif request == RPCStartupRequest.GET_SCHEDULER_CONFIG:
+                        response = self.engine.get_scheduler_config()
+                    elif request == RPCStartupRequest.GET_PARALLEL_CONFIG:
+                        response = self.engine.get_parallel_config()
+                    elif request == RPCStartupRequest.GET_TRACING_ENABLED:
+                        response = self.engine.is_tracing_enabled()
+                    elif request == RPCStartupRequest.IS_SERVER_READY:
+                        response = VLLM_RPC_SUCCESS_STR
+                    elif request == RPCStartupRequest.CLIENT_IS_READY:
+                        response = VLLM_RPC_SUCCESS_STR
+                        # Breakout of loop once client is ready.
+                        client_is_ready = True
+
+                    socket.send_multipart((identity, pickle.dumps(response)),
+                                          copy=False)
+
+                except Exception as e:
+                    socket.send_multipart((identity, pickle.dumps(e)),
+                                          copy=False)
+
+    def run_engine_loop(self) -> None:
```

Comment:
Is it possible and beneficial to use async functions for all the function calls in this function?

Reply:
What I gathered from this PR description is that this new engine implementation is running in a process without the use of any asyncio event loops, it's just a serial loop taking inputs over zmq and passing them to an `LLMEngine`

Reply:
no, there is no asyncio at all in this design

Reply:
I know. I'm just wondering whether making the following functions async could further help with performance:

```
self.maybe_handle_new_input()
request_outputs = self.engine.step()
self.stream_outputs(request_outputs)
```

Reply:
It is a good point. I have a follow up PR https://github.com/vllm-project/vllm/pull/8125 that makes maybe_handle_new_input() and stream_outputs() async with GPU forward pass and it indeed helps. Maybe we will integrate the changes there immediately into this PR, but not sure yet.

Reply:
I have added the async behavior of sockets to this PR (and closed 8125)